sphinx/cmd/quickstart.py

```diff
@@ -332,65 +352,42 @@ def ask_user(d):
         print()
-        do_prompt(d, 'master', 'Please enter a new file name, or rename the '
+        d['master'] = do_prompt('Please enter a new file name, or rename the '
                   'existing file and press Enter', d['master'])
 
-    if 'epub' not in d:
-        print('''
-Sphinx can also add configuration for epub output:''')
-        do_prompt(d, 'epub', 'Do you want to use the epub builder (y/n)',
-                  'n', boolean)
-
-    if 'ext_autodoc' not in d:
-        print('''
-Please indicate if you want to use one of the following Sphinx extensions:''')
-        do_prompt(d, 'ext_autodoc', 'autodoc: automatically insert docstrings '
-                  'from modules (y/n)', 'n', boolean)
-    if 'ext_doctest' not in d:
-        do_prompt(d, 'ext_doctest', 'doctest: automatically test code snippets '
-                  'in doctest blocks (y/n)', 'n', boolean)
-    if 'ext_intersphinx' not in d:
-        do_prompt(d, 'ext_intersphinx', 'intersphinx: link between Sphinx '
-                  'documentation of different projects (y/n)', 'n', boolean)
-    if 'ext_todo' not in d:
-        do_prompt(d, 'ext_todo', 'todo: write "todo" entries '
-                  'that can be shown or hidden on build (y/n)', 'n', boolean)
-    if 'ext_coverage' not in d:
-        do_prompt(d, 'ext_coverage', 'coverage: checks for documentation '
-                  'coverage (y/n)', 'n', boolean)
-    if 'ext_imgmath' not in d:
-        do_prompt(d, 'ext_imgmath', 'imgmath: include math, rendered '
-                  'as PNG or SVG images (y/n)', 'n', boolean)
-    if 'ext_mathjax' not in d:
-        do_prompt(d, 'ext_mathjax', 'mathjax: include math, rendered in the '
-                  'browser by MathJax (y/n)', 'n', boolean)
-    if d['ext_imgmath'] and d['ext_mathjax']:
-        print('''Note: imgmath and mathjax cannot be enabled at the same time.
-imgmath has been deselected.''')
-        d['ext_imgmath'] = False
-    if 'ext_ifconfig' not in d:
-        do_prompt(d, 'ext_ifconfig', 'ifconfig: conditional inclusion of '
-                  'content based on config values (y/n)', 'n', boolean)
-    if 'ext_viewcode' not in d:
-        do_prompt(d, 'ext_viewcode', 'viewcode: include links to the source '
-                  'code of documented Python objects (y/n)', 'n', boolean)
-    if 'ext_githubpages' not in d:
-        do_prompt(d, 'ext_githubpages', 'githubpages: create .nojekyll file '
-                  'to publish the document on GitHub pages (y/n)', 'n', boolean)
-
-    if 'no_makefile' in d:
-        d['makefile'] = False
-    elif 'makefile' not in d:
+    if 'builders' not in d:
+        print('Indicate which of the following builders configuration should '
+              'be included for:')
+        d['builders'] = []
+        for name, description in BUILDERS:
+            if do_prompt('%s: %s (y/n)' % (name, description), 'n', boolean):
+                d['builders'].append(name)
```

Comment:
Did you mean all builders in BUILDERS are disabled by default? Really?

Reply:
Not quite. Per one of the commit messages and the pull request summary:

> For some reason, we include the epub builder config from 'conf.py' in the list of "things users can modify", yet we don't do this for any other builder. Most users only care about the HTML builder and everything else is noise. As such, we should make this configurable and default all non-HTML builders to False, with the option to override this behavior.

As a result, the `html` builder is always enabled and other builders are disabled and most people don't use them. This is the same thing we do for extensions (enable them only if you actually want them)

Reply:
HTML builder is only a part of Sphinx. Other builders are also important for Sphinx. So I opposite for your proposal.

Reply:
Right, but I'm not removing anything: the users will still get asked if they want to use a builder or not (see below). As it is, the `conf.py` file already includes a lot of configuration options we really don't need. However, even with the current size of the file, we still don't include configuration for _extensions_ we aren't using. Why include configuration for _builders_ we aren't using? Surely if a user wanted to build PDFs or generate man pages then they would say so?

---

    htmlhelp: generate HTML help support files for Microsoft HTML Help Workshop (y/n)
    latex: generate LaTeX files that can be processed into PDF and other formats (y/n)
    man: generate man pages in groff format (y/n)
    texinfo: generate Texinfo files that can be processed into Info files (y/n)
    epub: generate an EPUB file for e-book readers (y/n)

Reply:
I agreed with current `conf.py` contains many options. And almost of them are commented out, so they are not important. But it does not means removing whole of them.
I believe some HTML options are also not needed.

I prefer to remove unneeded options and commented-out blocks from initial `conf.py`. It's acceptable for me.

Reply:
I'm not sure what you mean. It sounds like you agree with me, but think we should go further and remove a lot more by default? If this is the case, would this be an acceptable first step? I can focus on removing more of the noise, including some of the HTML builder options in another pull request? This pull request is focused on making builder configuration optional

Reply:
Sorry for confuse you. I said NO to make builder configuration optional.

Reply:
So what would you like me to do here? Drop the entirety of [the "configurable builder" change](https://github.com/sphinx-doc/sphinx/pull/4174/commits/2a709b1866557a1ae253d7a083cea7e20593afb4) or rework it to default every builder to enabled?

Reply:
I think dropping the commit is better.

Reply:
No. Personally, I proposed to reduce questions of quickstart. So your proposal conflicts with mine.

And we have to know the effect of the change. At least, we have to know how they works when all configurations are removed. And we should provide the way to enable them.

Are the options really important for users? I don't think so.

Reply:
So, let's make sure I'm getting this clear :smile: My problem is that this is all 90% of people need in their configuration file:

    master_doc = 'index'

That's it. If you're doing what most people do and using the `html` builder with no extensions to publish docs to readthedocs.org then nothing else is necessary. Compare this to the `conf.py` files that most projects have, like [this one](https://github.com/openstack/nova/blob/15.0.0/doc/source/conf.py). All those comments are noise and we don't need them. Instead we should simply be pointing people to docs if they want to configure stuff. This idea aims to tackle part of that by removing one thing that people almost certainly won't need: configuration for other builders.

I think you're suggesting that we ask ask *fewer questions* up front but have *more options* in the resulting `conf.py`? Meanwhile, I'm suggesting we ask *more questions* up front but have *fewer options** in the resulting `conf.py`. I think my option is a good one because you only run `sphinx-quickstart` once per project, whereas you might have to modify `conf.py` hundreds or thousands of times throughout a projects lifetime. In essence, the former approach is **lower initial cost** for **higher lifetime cost**. My proposal has **higher initial cost** for **lower lifetime cost**. Suggesting that users spend a little longer tweaking their `conf.py` is a reasonable ask. If they do not wish to do this then they can pass the ``--quiet`` option and ignore this whole process.

I really don't want to drop this because while you want a smaller number of options, I want a smaller `conf.py`. If you still don't agree with the above (and I urge you to think about it), what other options are available? Could I simply remove the `--epub` option and remove all non-HTML configuration options from `conf.py`? Maybe I could remove the HTML configuration options too? Maybe I could even go as far as removing *all* commented out options, and adding a link to the configuration guide in the comment at the top? Thoughts?

Reply:
I don't opposite to drop options from `conf.py`. I only opposite to make builders optional. We can reduce options without it.

Anyway, I feel very tired for reviewing your PR. So I'll stop maintainance work for a while.
Thanks,

Reply:
Oh, I'm sorry - I didn't mean to be rude and if I came across that way then I do apologize :disappointed: I'm trying to understand your objections so we can come to a compromise that makes everyone happy. This took a lot of my (personal) time and I don't want it to go to waste. I'm also trying my best to help review other patches and bugs and have a big documentation rework ready to go, to help you guys if possible. I'll drop this patch from the series for now, but I would like to address the whole "`conf.py` is too darn big" issue in the future. What would be the best way to discuss this?

Reply:
I'm back now. Sorry, I was very stupid. I must apologize you to my rudeness. I'd like to collaborate with you if you will forgive me.

To make a step forward, could you add a new issue to simplify conf.py please? I feel this PR contains a lot of elements; refactoring, interface for extensions and making builders optional. IMO, one issue one problem is better.

Reply:
It's no problem :+1: Maintaining stuff is stressful work for little reward, so totally understandable. I'm out of the country at the moment, but I'll take a look at this (and other comments) when I return in a week. Hopefully that's OK.

Reply:
@tk0miya This is noted in #4256. I have removed the offending patch here now. Is anything else needed?